Add New Layouts and Improve Navigation UI

jrakibi opened the pull request:
This PR introduces the following updates:


- [x] Adds a new layout for the project section
- [x] Implements a layout for the Bitcoin development history page
- [x] Creates a new component for mailing list signups
- [x] Updates the navigation UI to support expand/collapse functionality for elements

Emmanuel-Develops commented:
Awesome work Jamal!!
I'm breaking this review into 2 sections

User Review and Code Review.
User Review First, so that Code Review is done in one pass

## Mobile Review
### Bug: sidebar navigation
issue: Glitchy scrolling through side panel, screen recording attached

https://github.com/user-attachments/assets/fcd6550f-d3b6-4ce8-be34-b440b910008c

### Bug: Breadcrumb
issue: 
- The breadcrumb becomes sticky and then stops being sticky randomly.
- The breadcrumb top isn't aligned to the bottom nav, there is little space in-between and blur doesn't get applied there
Screen recording below

https://github.com/user-attachments/assets/ff90bd84-f86c-4cf8-b928-2c187d855ae8

### UX: Bitcoin History
issue: This isn't much of an issue but navigating through bitcoin history feels quirky.
suggestions: increase tolerance for the history events (stay longer on screen before considered not in view).
Also I think a one way animation is fine, i.e they pop in when scrolling to the bottom but they don't pop out. So when user scrolls back up the previously rendered ones are still visible. Similar to the `connectingLines`

https://github.com/user-attachments/assets/590d406f-6548-43bb-b801-35f896ff8769

Emmanuel-Develops commented:
### Bug: Breadcrumb
Issue: On larger mobile screens and in '/topics' route, the breadcrumb hovers close to the bottom.
Its present in mobile but only visible in large mobile screens. Looks like it is caused by the `banner` component. 

![Screenshot_20240910_114802_Chrome](https://github.com/user-attachments/assets/4a81e9bd-cf88-46dc-a50a-70cac2480de8)

jrakibi commented:
@Emmanuel-Develops great catch, the fix is in this commit https://github.com/bitcoin-dev-project/bitcoin-dev-project/pull/102/commits/b50137e44036a24ed39bdbb7cb6572218cbd654c

Emmanuel-Develops commented on components/bitcoin-topics/markdown-ui/bitcoin-history.tsx:
```diff
@@ -0,0 +1,587 @@
+    )
+}
+
+const bitcoinHistory: HistoryEvent[] = [
```
What are your thoughts regarding this data being available on `bitcoin-topics` repo instead.

currently suggest-edits action navigates the user to `https://github.com/jrakibi/bitcoin-topics/edit/main/topics/bitcoin-history.mdx` (I manually checked the feature/project branch for the file). And it has only the metadata and `<BitcoinHistory />` component.

My suggestion is the data should be passed from that mdx file e.g
```mdx
---
title: "Bitcoin History"
date: 2024-07-05T15:32:14Z
...
---

<BitcoinHistory data={[
  {
    id: "2007-post-satoshi-starts",
    type: "post",
    ...
  }
]}/>
```

What do you think?

jrakibi replied:
that's a great idea for separation of concerns. It allows us to update the data without affecting the `bitcoin-dev-project` repo

Emmanuel-Develops commented on components/bitcoin-topics/topic/TopicHeader.tsx:
```diff
@@ -1,16 +1,22 @@
 "use client"
 
+import { usePathname } from "next/navigation"
 import Metadata from "./metadata"
```
Metadata is unused

Emmanuel-Develops commented on components/bitcoin-topics/Navigation.tsx:
```diff
@@ -40,7 +47,23 @@ export function Navigation({
+
+    // Load expanded state from localStorage on component mount
+    useEffect(() => {
+        const savedState = localStorage.getItem("expandedTopics")
```
Why are we tracking and saving expanded sections?

I get the idea of resuming where the last stopped (good idea).
A couple of nits tho. 
- It doesn't track the exact topic they were on last.
- Multiple sections can be expanded as well, this doesn't provide much information.

Is there a specific need for the current implementation? If not we can remove it.

jrakibi replied:
Is there any reason we want to save the exact topic they were on last?

Having multiple expanded sections is beneficial ig, as it allows users to keep sections of interest open. This is also beneficial for users who are exploring or comparing different areas of content simultaneously

Emmanuel-Develops replied:
I was thinking in line of user progress, so they can continue from where they last stopped.

Bitcoin topics layout has a hierarchy/progression structure and since we do not track user progression it would be nice to keep track of the last topic they were on. At least that was why I thought you had that in there.

Seems your implementation serves a different purpose.
And user progress is another conversation.

 > Having multiple expanded sections is beneficial ig, as it allows users to keep sections of interest open. This is also beneficial for users who are exploring or comparing different areas of content simultaneously
 
 It does this already right? in state. The extra bit of localStorage is what I struggle to see the benefits, that is over different browser sessions for the user. 
Extra features do not hurt so if you find it valuable then kindly keep it as is 🙏

jrakibi replied:
Ooooh, I understand your point now! That's a great approach, appreciate your suggestion.

Working on it and will push the update soon.

<img width="343" alt="image" src="https://github.com/user-attachments/assets/85915d0f-68c9-4ba7-a66b-773063491f68">

Emmanuel-Develops commented on components/bitcoin-topics/Navigation.tsx:
```diff
@@ -40,7 +47,23 @@ export function Navigation({
+    useEffect(() => {
+        const savedState = localStorage.getItem("expandedTopics")
+        if (savedState) {
+            setExpandedTopics(JSON.parse(savedState))
```
parsing JSON should be in a try/catch block. There is always a chance to error out if it parses incorrect json formatted string. Given its pulling its data from local storage. The string cannot be trusted to always be properly formatted.

Emmanuel-Develops commented on components/bitcoin-topics/Navigation.tsx:
```diff
@@ -117,35 +179,66 @@ export function Navigation({
+                                            link.children.length > 0 && (
+                                                <span
+                                                    className={clsx(
+                                                        "ml-auto mr-3 text-xs flex-shrink-0",
```
the touch target to expand/collapse child sections is small on mobile and hard to press.
This leads to accidentally clicking the parent in most cases.
A padding around those icons could help

Emmanuel-Develops commented on components/bitcoin-topics/layouts/TopicBanner.tsx:
```diff
@@ -0,0 +1,265 @@
+import {
+    motion,
+    AnimatePresence,
+    useMotionValue,
```
the following variables are never used `useMotionValue, useSpring, PanInfo`

Emmanuel-Develops commented on components/bitcoin-topics/layouts/TopicLayout.tsx:
```diff
@@ -38,36 +42,26 @@ export default function TopicLayout({
-    const swipeXSmooth = useSpring(swipeX, { stiffness: 300, damping: 30 })
-    const swipeOpacity = useTransform(swipeXSmooth, [-50, 0, 50], [0.5, 0, 0.5])
+    const [currentPath, setCurrentPath] = useState<string[]>([])
+    const [expandedTopics, setExpandedTopics] = useState<Set<string>>(new Set())
```
`expandedTopics` is never used

Emmanuel-Develops commented on components/bitcoin-topics/layouts/TopicLayout.tsx:
```diff
@@ -38,36 +42,26 @@ export default function TopicLayout({
-        } else if (info.offset.x < -50 && isNavOpen) {
-            setIsNavOpen(false)
-        }
+    const toggleExpand = (href: string) => {
```
`toggleExpand` is never used

Emmanuel-Develops reviewed (changes requested):
Looks great, A couple of review comments

jrakibi commented:
Thanks @Emmanuel-Develops for the review

I addressed your feedback in this commit https://github.com/bitcoin-dev-project/bitcoin-dev-project/pull/102/commits/ddd0dcf855f166b216c1c557712d943671bd234c

I think the only thing left is the one related to navigation and expanded sections, as I believe there's value in keeping multiple sections of navigation open at the same time. I've left a comment here: https://github.com/bitcoin-dev-project/bitcoin-dev-project/pull/102#discussion_r1754190678

Emmanuel-Develops commented on components/bitcoin-topics/markdown-ui/hex-transaction-highlighter.tsx:
```diff
@@ -0,0 +1,284 @@
+                </div>
+                <div className="flex items-center justify-between mb-2">
+                    <div className="text-base">
+                        {gameState === "playing"
```
```suggestion
                        {gameState === GameState.Playing
```

Emmanuel-Develops commented on components/bitcoin-topics/markdown-ui/hex-transaction-highlighter.tsx:
```diff
@@ -0,0 +1,284 @@
+                            </motion.button>
+                        </motion.div>
+                    )}
+                    {gameState === "playing" && (
```
same her

Emmanuel-Develops commented on components/bitcoin-topics/markdown-ui/hex-transaction-highlighter.tsx:
```diff
@@ -0,0 +1,284 @@
+                            </motion.div>
+                        </motion.div>
+                    )}
+                    {gameState === "showingResult" && (
```
here as well

Emmanuel-Develops commented on components/bitcoin-topics/markdown-ui/hex-transaction-highlighter.tsx:
```diff
@@ -0,0 +1,284 @@
+
+            {/* Footer */}
+            <div className="bg-vscode-input-light dark:bg-vscode-input-dark p-4 flex justify-end">
+                {gameState === "playing" && (
```
also here

Emmanuel-Develops commented:
> Thanks @Emmanuel-Develops for the review
> 
> I addressed your feedback in this commit [ddd0dcf](https://github.com/bitcoin-dev-project/bitcoin-dev-project/commit/ddd0dcf855f166b216c1c557712d943671bd234c)
> 
> I think the only thing left is the one related to navigation and expanded sections, as I believe there's value in keeping multiple sections of navigation open at the same time. I've left a comment here: [#102 (comment)](https://github.com/bitcoin-dev-project/bitcoin-dev-project/pull/102#discussion_r1754190678)

Awesome! That was wicked quick.
I left a reply [here](https://github.com/bitcoin-dev-project/bitcoin-dev-project/pull/102#discussion_r1754253626)

Emmanuel-Develops commented:
Vercel build error
```bash
Error: Expected component `ExerciseSelector` to be defined: you likely forgot to import, pass, or provide it.
    at r (eval at getMDXComponent (/vercel/path0/.next/server/chunks/112.js:154:17720), <anonymous>:152:260)
    at m (eval at getMDXComponent (/vercel/path0/.next/server/chunks/112.js:154:17720), <anonymous>:3:1057)
    at u (eval at getMDXComponent (/vercel/path0/.next/server/chunks/112.js:154:17720), <anonymous>:152:233)
    at em (/vercel/path0/node_modules/next/dist/compiled/next-server/app-page.runtime.prod.js:76:10452)
    at /vercel/path0/node_modules/next/dist/compiled/next-server/app-page.runtime.prod.js:76:22148
    at Object.toJSON (/vercel/path0/node_modules/next/dist/compiled/next-server/app-page.runtime.prod.js:76:25726)
    at stringify (<anonymous>)
    at eE (/vercel/path0/node_modules/next/dist/compiled/next-server/app-page.runtime.prod.js:76:14115)
    at eR (/vercel/path0/node_modules/next/dist/compiled/next-server/app-page.runtime.prod.js:76:14558)
    at Timeout._onTimeout (/vercel/path0/node_modules/next/dist/compiled/next-server/app-page.runtime.prod.js:76:11338)
```

jrakibi commented:
@Emmanuel-Develops  The issue with Vercel deployment should be fixed now.

I also added a UI to display the last visited topic for the user based on our discussion here: https://github.com/bitcoin-dev-project/bitcoin-dev-project/pull/102#discussion_r1754253626

Here's the commit for the UI: https://github.com/bitcoin-dev-project/bitcoin-dev-project/pull/102/commits/d3ce23d18e10d9f36b13a4f3f9f7ffb83a27344a

jrakibi commented:
@IgboPharaoh this is the PR

the commits are :
- [d3ce23d](https://github.com/bitcoin-dev-project/bitcoin-dev-project/pull/102/commits/d3ce23d18e10d9f36b13a4f3f9f7ffb83a27344a) this add a UI that shows users the last visited topic where they left off
- [9ed80e1](https://github.com/bitcoin-dev-project/bitcoin-dev-project/pull/102/commits/9ed80e1a96ecfe84cc337edb2b47e815622ac0fd) the new UI for opcode explorer

Emmanuel-Develops commented:
# User Review
## Mobile Review
### Observations: OP_CODE EXPLORER
- clicking next button at the end of the animation replays the animation. See screen recording below

https://github.com/user-attachments/assets/7c50a735-d252-4de8-8185-03c47a765f56

- have you considered disabling buttons while the animation is playing? The UI suggests its disabled but they click regardless.
<img width="253" alt="Screenshot 2024-09-16 at 6 27 23 PM" src="https://github.com/user-attachments/assets/0d61ddff-8571-4ea8-881e-b99660303687">

- we can remove the toggle between compact and extended for mobile since it is always in extended view
<img width="256" alt="Screenshot 2024-09-16 at 6 26 59 PM" src="https://github.com/user-attachments/assets/0acc90fd-36ef-4f10-a38d-65a6297fdb0f">

jrakibi commented:
this commit https://github.com/bitcoin-dev-project/bitcoin-dev-project/pull/102/commits/0155481e545318187f5371a8e7c96496a56a389a addresses the issues you mentioned above 
- Disabled Previous/Next buttons at animation beginning/end of an animation
- Removed the toggle for mobile

IgboPharaoh commented:
Great work Jamal, here are some comments from a user review I did:

- Is there a reason why `Bonus: Problems with P2PK` is the next button  on the topics home page and it's not the next topic in the navigation setup
![Screenshot 2024-09-18 at 10 02 27](https://github.com/user-attachments/assets/01100db3-5227-4165-8059-3a993821a650)
- When I select a navigation item that is collapsible, why does the collapsible or caret icon stops working. Is this the intended functionality ?
![Screenshot 2024-09-18 at 09 58 33](https://github.com/user-attachments/assets/5fe3253e-1e86-4141-b90d-0c5cf17018bc)

IgboPharaoh commented on components/bitcoin-topics/markdown-ui/mailing-list-signup.tsx:
```diff
@@ -0,0 +1,162 @@
+        setLoading(true)
+
+        try {
+            const response = await fetch("/subscribe", {
```
You're calling a subscribe route but I can't find the route implementation.
Has this been implemented before ?

IgboPharaoh commented on components/bitcoin-topics/Navigation.tsx:
```diff
@@ -40,7 +48,55 @@ export function Navigation({
+            "expandedChildren",
+            JSON.stringify(expandedChildren)
+        )
+    }, [expandedTopics])
```
Github actions is suggesting adding 'navigation' as a dependency in the dependency array, you might want to check it out

jrakibi replied:
can you point me to the github action log

IgboPharaoh commented on components/bitcoin-topics/Navigation.tsx:
```diff
@@ -77,32 +133,95 @@ export function Navigation({
+    }, [pathname, navigation])
+
+    const getIcon = useCallback((iconName: string) => {
+        const IconComponent =
```
Interesting piece of code here, is this IconComponent, global ?

jrakibi replied:
this allows dynamic access to any icon from react-icons. It's not global, it's scoped within the getIcon function

IgboPharaoh replied:
Ohhh alright, since react icons can be imported anywhere I have no issues with it

Emmanuel-Develops commented on components/bitcoin-topics/markdown-ui/bitcoin-history.tsx:
```diff
@@ -0,0 +1,400 @@
+export const BitcoinHistory: React.FC<BitcoinHistoryProps> = ({
+    historyEvents
+}) => {
+    const [visibleEvents, setVisibleEvents] = useState(historyEvents.length)
```
this is not used

IgboPharaoh commented on components/bitcoin-topics/Navigation.tsx:
```diff
@@ -184,6 +347,43 @@ export function Navigation({
+
+    const handleSubmit = (e: React.FormEvent) => {
+        e.preventDefault()
+        console.log("Submitted:", email)
```
I think the console.log is not necessary, also this component is not connected to the subscribe route as it only clears the input value, please connect it

jrakibi replied:
good catch on that, impelemented the logic here https://github.com/bitcoin-dev-project/bitcoin-dev-project/pull/102/commits/7e434bd8855108b2f2ed031c64a2d6d231374c27

Emmanuel-Develops commented on components/bitcoin-topics/markdown-ui/opcode-explorer.tsx:
```diff
@@ -1,14 +1,16 @@
+import { ResetIcon } from "@radix-ui/react-icons"
+import { motion } from "framer-motion"
 import {
     SearchIcon,
```
unused import

IgboPharaoh commented on components/bitcoin-topics/hero/Hero.tsx:
```diff
@@ -34,182 +43,459 @@ function TrafficLightsIcon(props: React.ComponentPropsWithoutRef<"svg">) {
+                            className="relative overflow-hidden cursor-pointer group"
+                            whileTap={{ scale: 0.98 }}
+                            onClick={() =>
+                                window.open(
```
Would an <a> </a> or <Link></Link> tag work in this scenario rather than using the global window object ?
If it would please replace

jrakibi replied:
fixed here https://github.com/bitcoin-dev-project/bitcoin-dev-project/pull/102/commits/f9a0cdc61d5166831306b155dddc90a4b186f7b5

IgboPharaoh commented on components/bitcoin-topics/layouts/TopicBanner.tsx:
```diff
@@ -0,0 +1,259 @@
+    const githubEditUrl = `https://github.com/jrakibi/bitcoin-topics/edit/main/topics/${content.slug}.mdx`
+
+    const EditOnGitHubButton = () => (
+        <a
```
Is there any reason you use the <a></a> over the next/Link tag ?

jrakibi replied:
Generally I use the `a` tag for external links, since there are no benefits to using `Link`, such as client-side routing, unless our concern is maintaining consistency by using the same tag throughout the application

IgboPharaoh commented on components/bitcoin-topics/layouts/TopicLayout.tsx:
```diff
@@ -82,14 +55,41 @@ export default function TopicLayout({
         </a>
```
Same as previous review comment on `a` tags

IgboPharaoh commented on components/bitcoin-topics/layouts/TopicLayout.tsx:
```diff
@@ -82,14 +55,41 @@ export default function TopicLayout({
     )
 
+    useEffect(() => {
+        if (isNavOpen) {
```
We're using similar piece of code in the `Topics Layout` and `Topics Banner` can this be abstracted into reusable functions and use in both places ? If it's not possible or will be time consuming please ignore for now

Emmanuel-Develops commented on components/bitcoin-topics/topic/TopicHeader.tsx:
```diff
@@ -1,16 +1,21 @@
 
 export function TopicHeader({ title, tags, summary }: any) {
+    const pathname = usePathname()
+    const isBitcoinHistory = pathname === "/topics/bitcoin-history"
```
There should be a better way for a topic to opt out of this rendering without having it explicitly defined here.
A way to do that could be to have a `renderTitle` field in the frontmatter of the topic.
But that means a change to all the topics markdowns and for such a small nit I don't think its advisable.

We can think about things like this going forward.

jrakibi replied:
You're right, this isn't ideal
for now, it's not worth changing all markdown files for this ig

Emmanuel-Develops commented on components/bitcoin-topics/Navigation.tsx:
```diff
@@ -9,12 +9,18 @@ import {
+import { motion } from "framer-motion"
+import { FaChevronRight, FaChevronDown } from "react-icons/fa"
+import * as Icons from "react-icons/fa"
+import { BookOpenIcon, ClockIcon, ArrowRightIcon } from "lucide-react"
```
these imports aren't used

Emmanuel-Develops commented on components/bitcoin-topics/Navigation.tsx:
```diff
@@ -40,7 +48,55 @@ export function Navigation({
+    const [expandedTopics, setExpandedTopics] = useState<
+        Record<string, boolean>
+    >({})
+    const [lastVisitedTopic, setLastVisitedTopic] = useState<string | null>(
```
lastVisitedTopic is not used.

jrakibi replied:
removed here https://github.com/bitcoin-dev-project/bitcoin-dev-project/pull/102/commits/16c40f9445f3a96c83a9e54b8e91235136e27a20

Emmanuel-Develops reviewed (changes requested):
Left some comments.
Looks good! only a few nits

jrakibi commented on components/bitcoin-topics/Navigation.tsx:
```diff
@@ -156,17 +311,25 @@ export function Navigation({
+                                                                    e
+                                                                ) => {
+                                                                    e.stopPropagation()
+                                                                    handleTopicClick(
```
`e.stopPropagation()` is used throughout the component to prevent clicks on the arrow from triggering the link click event. But Ithink it may not be needed in this speicific case. 
Removed here https://github.com/bitcoin-dev-project/bitcoin-dev-project/pull/102/commits/bb01fb7005ae14728d77f6afa66d485c56b6d766

jrakibi commented:
Thanks @IgboPharaoh for the review, I believe I adressed all your feedback

For the remaining comments:

- `Bonus: Problems with P2PK` being the next button on the topics home page is wrong behaviour, it's fixed now

- No changes are needed here as we expect an active link (parent or child) to remain expandable even after clicking the chevron icon




> Great work Jamal, here are some comments from a user review I did:
> 
> * Is there a reason why `Bonus: Problems with P2PK` is the next button  on the topics home page and it's not the next topic in the navigation setup
>   ![Screenshot 2024-09-18 at 10 02 27](https://private-user-images.githubusercontent.com/87058633/368496006-01100db3-5227-4165-8059-3a993821a650.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjY2NTc3NjAsIm5iZiI6MTcyNjY1NzQ2MCwicGF0aCI6Ii84NzA1ODYzMy8zNjg0OTYwMDYtMDExMDBkYjMtNTIyNy00MTY1LTgwNTktM2E5OTM4MjFhNjUwLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA5MTglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwOTE4VDExMDQyMFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWY1MDc0NzEyYTVhZWNmNzJiZDI1ZTYzZDViYzk2NGY3OGI2NGRlNzQxMWY1YmMyNzcyZWFlYzM1NzA5NzA0NTImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.anlg032WuAVwZHXaIrNnXaA5incN1N4oaxiWJppFsMY)
> * When I select a navigation item that is collapsible, why does the collapsible or caret icon stops working. Is this the intended functionality ?
>   ![Screenshot 2024-09-18 at 09 58 33](https://private-user-images.githubusercontent.com/87058633/368497081-5fe3253e-1e86-4141-b90d-0c5cf17018bc.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjY2NTc3NjAsIm5iZiI6MTcyNjY1NzQ2MCwicGF0aCI6Ii84NzA1ODYzMy8zNjg0OTcwODEtNWZlMzI1M2UtMWU4Ni00MTQxLWI5MGQtMGM1Y2YxNzAxOGJjLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA5MTglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwOTE4VDExMDQyMFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTJjYjMyYTIzZWNlODM2NGNjZjhkNzRlMjc0ZTY3MjUwZjRiNzg0MzhiM2VmNWZkMGU1YmNkMTdhOGExZDBiNTMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.DHDoFMbcUlaJTthD3odCDDKGX78qR73r-8e6eHpGNco)

jrakibi commented on components/bitcoin-topics/Navigation.tsx:
```diff
@@ -77,32 +133,95 @@ export function Navigation({
+        return <IconComponent className="inline-block mr-2 text-current" />
+    }, [])
+
+    const handleTopicClick = useCallback(
```
That’s a great catch, thanks, It’s fixed here: https://github.com/bitcoin-dev-project/bitcoin-dev-project/pull/102/commits/9b7bf29fea7c2f9057e6e607f6a4f08d2b6b3fa5.

For the excluded routes, let’s keep them as they are for now. We’ll implement tracking of which topics the student has read in the next release

jrakibi commented on components/bitcoin-topics/layouts/TopicBanner.tsx:
```diff
@@ -0,0 +1,259 @@
+"use client"
```
same reply as https://github.com/bitcoin-dev-project/bitcoin-dev-project/pull/102#discussion_r1764924235

jrakibi commented on components/bitcoin-topics/hero/Hero.tsx:
```diff
@@ -34,182 +43,459 @@ function TrafficLightsIcon(props: React.ComponentPropsWithoutRef<"svg">) {
+                                </motion.div>
+                            </div>
+                            <motion.div
+                                className="absolute inset-0 bg-indigo-100 dark:bg-indigo-900 opacity-0"
```
removed

jrakibi commented:
@Emmanuel-Develops I’ve added inline comments. 
I’ve addressed all feedback except the one about the `TopicBanner` and `TopicLayout` abstraction. I left a reference comment for my reply to Solomon regarding that question

jrakibi commented:
@Emmanuel-Develops I removed saving the `time` from local storage as suggested by Jonas yesterday https://github.com/bitcoin-dev-project/bitcoin-dev-project/pull/102/commits/47192dfc374eba3b1d80854091a8ed5876fa8ef8

jrakibi commented on components/bitcoin-topics/topic/PrevNextLinks.tsx:
```diff
@@ -54,24 +62,34 @@ export function PrevNextLinks({
                     className="ml-auto text-right"
                     title={prev.title}
                     href={`/${prev.path}`}
+                    onClick={() => handleTopicClick(`/${prev.path}`)}
```
@Emmanuel-Develops fixed here https://github.com/bitcoin-dev-project/bitcoin-dev-project/pull/102/commits/8e89cdc2d565d38d9376377b60544ac7a1c4a3eb
